### DISCUSSION TITLE
[Application Development] How to find the closest L1 block number from L2 block number?

### DISCUSSION THREAD

**Discussion by bap2pecs** [ethereum-optimism/developers#443] [Application Development] [answered]
### Did you check the documentation?

- [X] Yes

### Did you check for duplicate questions?

- [X] Yes

### Issue Description

If I know the block number of an OP-stack L2, can I calculate the latest L1 block number at the time this block was generated? I can look up the block closest in time using the timestamp, but I want to know if there's an easier method (e.g., using epoch number or something similar).

### Additional Information

in our application use case, we need to query some oracle contract deployed on L1 when receiving a L2 block. That's why we want to know the L1 block.

we can use "latest" block tag when querying the RPC but there is another use case where we want to query such data from an old L2 block. that's why I have this questions

**Comment by opfocus** [comment, marked as answer]
CC @sbvegan 
I would like to share a preliminary observation. By examining the input data of this L2 transaction: https://sepolia-optimism.etherscan.io/tx/0x92176aca750cf02d0e23db9bb66fa941d702470d5d1d3240434b3f6f579d7685, it appears to contain information about the L1 block. The timestamps of these two blocks are very close, and I believe that every L2 block includes such a transaction

**Reply by bap2pecs** [reply]
thanks @opfocus! I think you are right. I also looked into the OP mainnet. for each OP L2 block, the first tx is always to set L1 block value which contains the L1 block number.

for example, L2 block [120924621](https://optimistic.etherscan.io/block/120924621) contains https://optimistic.etherscan.io/tx/0xf78fd113dbb7c75171b3b42bf586c8d6de82af8aec3a9acdf339b36ca2694eb3 with calldata:

> 0x440a5e2000000558000c5fc5000000000000000000000000665e2ccb00000000013162e1000000000000000000000000000000000000000000000000000000026a5198bc00000000000000000000000000000000000000000000000000000000000000013b36ff064300c712d0b12a984b44ac83cf8abaa938989397b39e62ab2c30f4310000000000000000000000006887246668a3b87f54deb3b94ba47a6f63f32985

per https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/L2/L1Block.sol/#L136, the timestamp and block number is from `sstore(number.slot, shr(128, calldataload(20)))`

the block number is from the second part: `00000000013162e1`, which is L1 block [20013793](https://etherscan.io/block/20013793)

L2 block 120924621 timestamp: Jun-03-2024 08:53:39 PM +UTC
L1 block 20013793 timestamp: Jun-03-2024 08:51:23 PM +UTC

which is very close

reference:
- https://ethereum.stackexchange.com/questions/67758/can-anyone-help-to-explain-calldataload
- https://ethereum.stackexchange.com/questions/113204/solidity-shr-opcode

**Reply by sbvegan** [reply]
Yep every L2 block has this system transaction, this is how the rollup gets L1 data and derives deposit transactions

**Reply by bap2pecs** [reply]
thanks @sbvegan! where is this part described in https://specs.optimism.io/? is it the same concept of "L1 origin"?

**Reply by lesterli** [reply]
yes, in the spec:

> in the L2 block, the first transaction MUST be a [L1 attributes deposited transaction](https://specs.optimism.io/protocol/deposits.html#l1-attributes-deposited-transaction).
